### ISSUE TITLE
Error Loading frozen graph(.pb) using readNetFromTensorflow 

### ISSUE THREAD

**Issue by PiotrSobczak** [opencv/opencv#14073]
##### System information (version)

- OpenCV => 4.0.0
- Operating System / Platform => Ubuntu 18.4
- Compiler => gcc 7.2

##### Detailed description

I'm having problems trying to load a model in pb or pb&pbtxt using readNetFromTensorflow. 
The model is a pretrained resnext network trained and exported using tflearn with code:
```
net = tflearn.input_data(shape=[None, 49, 49, 3], img_prep, img_aug)
net = tflearn.conv_2d(net, 16, 3, regularizer='L2', weight_decay=0.0001)
...
net = tflearn.fully_connected(net, 11, activation='softmax')

model = tflearn.DNN(net)
model.load("checkpoints/meta-data/model_resnext-96000")

init_op = tf.initialize_all_variables()
sess = tf.Session()
sess.run(init_op)

tf.train.write_graph(sess.graph_def, "/tmp/my-model", 'graph.pbtxt', as_text=True)
tf.train.write_graph(sess.graph, "/tmp/my-model", 'graph.pb', as_text=False)
```
The model works perfectly fine using tflearn api (model.predict()). 
After that I attempted to freeze the graph and optimize it for inference:

1.Freezing the graph
```
python -m tensorflow.python.tools.freeze_graph --input_binary --input_graph graph.pb --output_graph frozen_graph.pb --input_checkpoint ~/datasets/TrainedNetworksTfLearn/Expression/model-resnext-no-train-ops --output_node_names FullyConnected/Softmax
```
2 Optimizing the graph for inference with two different ways
a).using tensorflow.python.tools.optimize_for_inference:
```
python -m tensorflow.python.tools.optimize_for_inference --input frozen_graph_with_freeze_graph_tool.pb --output frozen_interference_graph_opt.pb --frozen_graph True --input_names InputData/X --output_names FullyConnected/Softmax
```

b). using snippet provided by @dkurt 

```
import tensorflow as tf
from tensorflow.tools.graph_transforms import TransformGraph
from tensorflow.python.tools import optimize_for_inference_lib

graph = 'graph.pb'
with tf.gfile.FastGFile(graph, "rb") as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    tf.summary.FileWriter('logs', graph_def)

    inp_node = 'InputData/X'
    out_node = 'FullyConnected/Softmax'
    graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def, [inp_node], [out_node],
                                                                  tf.float32.as_datatype_enum)
    graph_def = TransformGraph(graph_def, [inp_node], [out_node], ["sort_by_execution_order"])

    tf.train.write_graph(graph_def, "/tmp/my-model", 'graph_opt.pbtxt', as_text=True)
    tf.train.write_graph(graph_def, "/tmp/my-model", 'graph_opt.pb', as_text=False)
```
This completes successfully:
>WARNING:tensorflow:From optimize_graph.py:6: FastGFile.__init__ (from tensorflow.python.platform.gfile) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.gfile.GFile.
WARNING:tensorflow:Passing a `GraphDef` to the SummaryWriter is deprecated. Pass a `Graph` object instead, such as `sess.graph`.
WARNING:tensorflow:From //.local/lib/python3.6/site-packages/tensorflow/python/tools/strip_unused_lib.py:86: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.extract_sub_graph
WARNING:tensorflow:From //.local/lib/python3.6/site-packages/tensorflow/python/tools/optimize_for_inference_lib.py:113: remove_training_nodes (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.remove_training_nodes
2019-03-15 17:32:23.963903: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying sort_by_execution_order

After these steps i get the following errors:
```
this->net = cv::dnn::readNetFromTensorflow("graph_opt.pb", "graph_opt.pbtxt");
```
> tf_importer.cpp:616: error: (-215:Assertion failed) const_layers.insert(std::make_pair(name, li)).second in function 'addConstNodes'
```
this->net = cv::dnn::readNetFromTensorflow("graph_opt.pb");
```
>  tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent'

**Comment by dkurt** [comment]
@PiotrSobczak,

```
net = tflearn.input_data(shape=[None, 49, 49, 3], img_prep, img_aug)
net = tflearn.conv_2d(net, 16, 3, regularizer='L2', weight_decay=0.0001)
...
net = tflearn.fully_connected(net, 11, activation='softmax')
```

Please provide a fully reproducible sample or attach all the mentioned file `graph_opt.pb`.

> OpenCV => 4.0.0

Have you tried to use newer version of OpenCV?

**Comment by PiotrSobczak** [comment]
@dkurt Thanks for quick response!. 
Full code in tflearn:
```
from __future__ import division, print_function, absolute_import
import tensorflow as tf
import tflearn

n = 5

# # Real-time data preprocessing
img_prep = tflearn.ImagePreprocessing()
img_prep.add_featurewise_zero_center(per_channel=True, mean=[0.53990436, 0.4405486, 0.39328504])

# Real-time data augmentation
img_aug = tflearn.ImageAugmentation()
img_aug.add_random_flip_leftright()
img_aug.add_random_crop([49, 49], padding=4)

# Building Residual Network
net = tflearn.input_data(shape=[None, 49, 49, 3], data_preprocessing=img_prep, data_augmentation=img_aug)
net = tflearn.conv_2d(net, 16, 3, regularizer='L2', weight_decay=0.0001)
net = tflearn.resnext_block(net, n, 16, 32)
net = tflearn.resnext_block(net, 1, 32, 32, downsample=True)
net = tflearn.resnext_block(net, n-1, 32, 32)
net = tflearn.resnext_block(net, 1, 64, 32, downsample=True)
net = tflearn.resnext_block(net, n-1, 64, 32)
net = tflearn.batch_normalization(net)
net = tflearn.activation(net, 'relu')
net = tflearn.global_avg_pool(net)

# Regression
net = tflearn.fully_connected(net, 11, activation='softmax')

# Initializing session
init_op = tf.initialize_all_variables()
sess = tf.Session()

# Training
model = tflearn.DNN(net, checkpoint_path='Snapshots/model_resnext',
                    max_checkpoints=10, tensorboard_verbose=0, tensorboard_dir='Logs/',
                    clip_gradients=0., session=sess)

sess.run(init_op)

# Saving binary and text graphs
tf.train.write_graph(sess.graph_def, "/tmp/resnext-model", 'graph.pbtxt', as_text=True)
tf.train.write_graph(sess.graph, "/tmp/resnext-model", 'graph.pb', as_text=False)
```
Loading this models causes:
```
this->net = cv::dnn::readNetFromTensorflow("/tmp/resnext-model/graph.pb");
```
> tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent
```
this->net = cv::dnn::readNetFromTensorflow(
    "/tmp/resnext-model/graph.pb", "/tmp/resnext-model/graph.pbtxt"
);
```
> tf_importer.cpp:530: error: (-2:Unspecified error) Const input blob for weights not found in function 'getConstBlob'

Then after applying your script for inference optimization I get slightly different error:
```
this->net = cv::dnn::readNetFromTensorflow("/tmp/resnext-model/graph_opt.pb");
```
>tf_importer.cpp:530: error: (-2:Unspecified error) Const input blob for weights not found in function 'getConstBlob'
```
this->net = cv::dnn::readNetFromTensorflow(
    "/tmp/resnext-model/graph_opt.pb", "/tmp/resnext-model/graph_opt.pbtxt"
);
```
> tf_importer.cpp:530: error: (-2:Unspecified error) Const input blob for weights not found in function 'getConstBlob'

**Comment by PiotrSobczak** [comment]
Upgrading to OpenCV(4.1.0-pre) didn't make any difference.

**Comment by gunnxx** [comment]
Have you tried to check its data type after each layer i.e. `tf.dtypes`? Seems like current supported Tensor's data type is limited to `DT_FLOAT`, `DT_DOUBLE`, `DT_INT32`, `DT_HALF`, and `DT_QUINT8`.

**Comment by PiotrSobczak** [comment]
I'm going to check it, but the after optimizing graph for inference the error changes from 
> tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent

to

> tf_importer.cpp:530: error: (-2:Unspecified error) Const input blob for weights not found in function 'getConstBlob'

**Comment by PiotrSobczak** [comment]
@dkurt Upgrading OpenCV didn't help. Any other ideas?

**Comment by dkurt** [comment]
@PiotrSobczak, Tha thing is that your model is not in the deployment phase. Check it's graph in TensorBoard:
![image](https://user-images.githubusercontent.com/25801568/54905008-fb7e8080-4ef0-11e9-9ecc-c3f96f1a81c8.png)


You need to freeze it.

**Comment by dkurt** [comment]
@PiotrSobczak, May I ask you to try a workaround from https://github.com/opencv/opencv/issues/14224#issuecomment-479021213?

**Comment by PiotrSobczak** [comment]
@dkurt Sure, I will give it a try by the end of the week.

**Comment by dkurt** [comment]
@PiotrSobczak, with optimizations from https://github.com/opencv/opencv/issues/14073#issuecomment-476098443 and the changes from https://github.com/opencv/opencv/pull/14251 I could import and launch your network.

**Comment by tobybreckon** [comment]
I have an identical problem, with a model I am trying to convert from tflearn and load into OpenCV via a frozen graph .pb file. Steps to reproduce my error are: 

```
git clone https://github.com/tobybreckon/fire-detection-cnn.git
cd fire-detection-cnn
git checkout convert-pb
sh ./download-models.sh    # which downloads models in tflearn checkpoint format
python pb-test.py          # which performs the "You need to freeze it" step described above
```

My resulting error on loading into opencv is:

```
  File "./pb-test.py", line 85, in <module>
    tensorflowNet = cv2.dnn.readNetFromTensorflow('minimal_graph.pb', 'minimal_graph.txt');
cv2.error: OpenCV(4.1.0-pre) /tmp/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:616: error: (-215:Assertion failed) const_layers.insert(std::make_pair(name, li)).second in function 'addConstNodes'

```

I also tried a variant on the various on-line advice on this:

```
python pb-test2.py 
```

My resulting error on loading into opencv is:
```
  File "./pb-test2.py", line 111, in <module>
    tensorflowNet = cv2.dnn.readNetFromTensorflow('frozen_inference_graph_opt.pb');
cv2.error: OpenCV(4.1.0-pre) /tmp/opencv/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent'


```

Any help appreciated. 

OpenCV: 4.1.0-pre

**Comment by dkurt** [comment]
@tobybreckon, please upload `frozen_inference_graph_opt.pb` somewhere and put a reference to download it. This way we can help you faster. Moreover there is no `pb-test.py ` in that branch anymore.

**Comment by tobybreckon** [comment]
@dkurt, thanks. It appears this issue is now also fixed by the latest #14251 fixes and pull into master - I tested last night, removing ```pb-test.py``` in the process. Many thanks.

**Comment by TanFluent** [comment]
@dkurt @PiotrSobczak thanks for your helps in this issues. And i manage to deploy faceNet .pb models on cv-dnn with try and error. I put my work [here](https://github.com/TanFluent/facenet_opencv_dnn). I think it may do a litlle help for those runing in trouble. LOL
